src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs

```diff
@@ -49,68 +54,85 @@ protected XDocument ToFlatOpcDocument(XProcessingInstruction instruction)
             // to call it here.
             Save();
 
+            // Identify all AlternativeFormatInputParts (AltChunk parts).
+            // This is necessary because AltChunk parts must be treated as binary
+            // parts regardless of the actual content type, which might even be
+            // XML-related such as application/xhtml+xml.
+            var altChunkPartUris = new HashSet<Uri>(
+                new InternalPackageRelationships(Package)
+                    .Where(pr => pr.RelationshipType == AltChunkRelationshipType)
+                    .Select(pr => PackUriHelper.ResolvePartUri(pr.SourceUri, pr.TargetUri)));
+
             // Create an XML document with a standalone declaration, processing
             // instruction (if not null), and a package root element with a
             // namespace declaration and one child element for each part.
             return new XDocument(
                 new XDeclaration("1.0", "UTF-8", "yes"),
                 instruction,
                 new XElement(
-                    pkg + "package",
-                    new XAttribute(XNamespace.Xmlns + "pkg", pkg.ToString()),
-                    Package.GetParts().Select(part => GetContentsAsXml(part))));
+                    Pkg + "package",
+                    new XAttribute(XNamespace.Xmlns + "pkg", Pkg.ToString()),
+                    Package.GetParts().Select(part => GetContentsAsXml(part, altChunkPartUris))));
         }
 
         /// <summary>
-        /// Gets the <see cref="PackagePart"/>'s contents as an <see cref="XElement"/>.
+        /// Gets the <see cref="PackagePart"/>'s XML or binary contents as an <see cref="XElement"/>.
         /// </summary>
         /// <param name="part">The package part.</param>
+        /// <param name="altChunkPartUris">The collection of AlternativeFormatInputPart URIs.</param>
         /// <returns>The corresponding <see cref="XElement"/>.</returns>
-        private static XElement GetContentsAsXml(PackagePart part)
+        private static XElement GetContentsAsXml(PackagePart part, IEnumerable<Uri> altChunkPartUris)
```

Comment:
The formal parameter does not have to be a `HashSet<T>` to utilize the O(1) for direct lookup. While it turns out that an `IEnumerable<T>` is not the optimal choice, because `Contains(T)` is an extension method rather than a method defined by the interface, the `Contains(T)` implementation of `HashSet<T>` will be used when it is passed as an actual parameter.

Having said that, I also think it would be better to use `HashSet<T>` or at a minimum `ICollection<T>`, which defines the `Contains(T)` method for `HashSet<T>`.